File: Day-4/day4.py
partA = 0
partB = 0

file = open("Day-4/input.txt", "r")

for line in file:
    curr = line.strip().split(",")
    firstRange = curr[0].split("-") 
    secondRange = curr[1].split("-") 

    if ((int(firstRange[0]) <= int(secondRange[0]) and int(firstRange[1]) >= int(secondRange[1])) or 
        (int(firstRange[0]) >= int(secondRange[0]) and int(firstRange[1]) <= int(secondRange[1]))):
        partA +=1

    if ((int(firstRange[0]) <= int(secondRange[0]) <= int(firstRange[1])) or 
        (int(secondRange[0]) <= int(firstRange[0]) <= int(secondRange[1]))):
        partB +=1

file.close()

print("Solution to part 1: ", partA)
print("Solution to part 2: ", partB)

File: Day-4/day4.hs
import Data.List 

main = do 
        fileData <- readFile "input.txt"
        let assignments = lines fileData
        let parsedAssignments = map (stringToInteger . (\(a,b) -> (splitListAt '-' a, splitListAt '-' b)) . (splitListAt ',')) assignments
        let partA = sum $ map (fromEnum . assignmentContains) parsedAssignments
        let partB = sum $ map (fromEnum . assignmentOverlaps) parsedAssignments
        print $ "Solution to part 1: " ++ show partA
        print $ "Solution to part 2: " ++ show partB

stringToInteger :: ((String, String), (String, String)) -> ((Int, Int), (Int, Int))
stringToInteger ((a, b), (x, y)) = ((read a :: Int, read b :: Int), (read x :: Int, read y :: Int))

splitListAt :: (Eq a) => a -> [a] -> ([a], [a])
splitListAt n xs = (\(x, y) -> (x, drop 1 y)) $ break (== n) xs 

assignmentOverlaps :: (Ord a) => ((a, a), (a, a)) -> Bool 
assignmentOverlaps ((a, b), (x, y)) = (a <= x && b >= x) || (x <= a && y >= a)

assignmentContains :: (Ord a) => ((a, a), (a, a)) -> Bool 
assignmentContains ((a, b), (x, y)) = (a >= x && b <= y) || (x >= a && y <= b)